### ISSUE TITLE
Edit Lock for Safe Simultaneous Edits?

### ISSUE THREAD

**Issue by karger** [mavoweb/mavo#727]
In multi-user apps that expect infrequent changes, it can still wreak havoc when 2 mavo authors just happen to edit and save at the same time.  A general purpose hack might be that clicking the "edit" button could check and then lock a semaphore in the data.  There could be a timeout or "break the lock" option on the semaphore to prevent permanent locking.

To avoid hard-coding, we could add an attribute `mv-edit-lock="propertyname"` that would tell mavo under what property it should use to store the semaphone information.

This isn't theoretical; I got this issue from discussion with a current mavo user who is unhappy with past save collisions in her app.

**Comment by LeaVerou** [comment]
Wouldn't it be a better solution to pull the most recent version of the data before saving, merge, then save with the new change?

**Comment by DmitrySharabin** [comment]
> Wouldn't it be a better solution to pull the most recent version of the data before saving, merge, then save with the new change?

I was thinking of using this approach in the GSheets plugin to benefit from the built-in changes history. I wonder wouldn’t it be too expensive from the number of network requests point of view?

**Comment by karger** [comment]
"merge" is such a little word for such a complicated concept.   given that the data can change in arbitrary ways, how do we safely perform a merge?   also, what do we do if there is a conflict at the most basic level, e.g. both users change the same number at the same time?

**Comment by LeaVerou** [comment]
What if we use more granular locks? Though then in the general case, we'd need to be able to *write* to the data just to put the lock in, which in itself creates a merge problem, doesn't it?

**Comment by karger** [comment]
indeed, there's no test-and-set primitive in the storage layer, so you'd have to store lock data in the data file.  that doesn't really create a merge problem though, because you'd be wanting to take the lock before you started editing.  there's no point in locking at the moment you want to save---that's too late to protect you from editing data that someone else is also editing.  all that's necessary is for you to notice quickly if someone else overwrote your lock.  e.g., if your using firebase with the realtime option and write a lock, and then someone else does, then you'll get the update notification which shows that you've lost the lock, and can react accordingly.

**Comment by LeaVerou** [comment]
This sounds very specific to Firebase. E.g. on Github or Dropbox this would pollute the version history, making it unusable.

**Comment by fregante** [comment]
I'm also facing issues with content being overwritten constantly. I tried instructing the user to refresh the page before loading but either that's not working or they forget to do it.

For me the solution would be to regularly fetch the data from GitHub and then either  warn the user that there's a new version so they don't start editing, or diff it with local and apply the changes immediately.

**Comment by LeaVerou** [comment]
Addressing the case of a user forgetting about the tab is a _much_ easier problem to solve than the case of two users editing simultaneously.

We could add an attribute that tells Mavo to fetch data periodically. It could be a plugin, at least at first. The backend would need to store the fetched data pre-parsing and compare it with the newly fetched version. If there are changes, then:

- if there are no unsaved changes, it just renders the data
- if there are unsaved changes, it asks the user what to do (which can be just a `confirm()` at first)
	- if the user opts *not* to load the new data, the periodic checking stops

This needs to happen in the background, so no `[mv-progress]` state change.
If `render()` can be called like this we should also make sure to restore state in it as much as possible, e.g. refocus focused elements if within a collection and things like that.

How does that sound?

> or diff it with local and apply the changes immediately.

The problem with that is that diffing can produce unreliable results.
